### PR TITLE
tools: Document and enforce well-known deps

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
We don't want code review to nitpick whether or not `@fmt` or `@eigen` is required in a BUILD file, because they are approximately _always_ available.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/12698)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
+@SeanCurtis-TRI could I get your initial take on this?  Is it a good idea?

**Comment by jwnimmer-tri** [comment]
I was definitely looking for explosive thoughts.  Let's see if my reasoning makes sense (and is documented well enough).

(1)

You're correct that we don't have a de jure style guide rule that developers need to deps-what-you use (DWYU).  In fact, the most closely related text is the [IWYU policy for includes](https://drake.mit.edu/styleguide/cppguide.html#Names_and_Order_of_Includes), where we write:

> While we encourage you to Include What You Use (IWYU) since it improves code transparency and readability, it will not be strictly enforced. Instead, we enforce a “weak include” style that simply requires every symbol referenced within a file be covered by the transitive closure of all #include statements in the file. We decided to make this exception because [ ... snip ... ] we don’t have a tool to enforce IWYU at this time. 

It is not unheard-of for a reviewer to comment that some `deps` are missing from a BUILD file during PR review, nor is it unheard-of for authors to scrub their deps prior to PR submission on a best-effort basis.  Both are doing so with good intentions -- so that unrelated future PRs don't trigger random-seeming breaks later.

I suspect what is happening is along the same line as "We encourage .." in the above text.  The goal of having robust source code is a good one, but without automated linting we stick with best-effort attempts to hit 90% compliance, instead of reading 100% compliance through tooling.

So, even though DWYU is not a de-jure defect, I believe we're still spending mental energy on it from time to time.  My hypothesis is that this PR would reduce the net energy spent.

As you say -- one solution to this is might be to discourage people from taking even a best-effort approach.  They should literally never think about it.  If the code builds, ship it.

In practice, I suspect that would lead to enough bitrot with missing deps that we'd have too many cases of removing a deps in one place uncovering it missing from a cascade of others.  When that fallout is far away from the new change, it costs more to fix than it would have cost to get it right during initial development and review, where the relevant code is under active study, paged in, etc.

TL;DR: I think people _do_ care about DWYU somewhat.

(2)

I  believe these two cases are special, because it is so unclear to authors when they need these or not.

So much code in Drake does `#include <Eigen...>` without a mention of `@eigen`, that when a PR comes up that _does_ say `@eigen`, it's a speed bump for a reader.  Why is this code special?  Are we using some special aspect of Eigen?  Should I go through my own code and say `@eigen`.

Similarly for fmt -- with pending standardization, and use almost everywhere in Drake it's rare to ever need to say `@fmt` in deps, so any times it is mentioned it stands out.

Given the consensus seems to be to _not_ declare these, but rather inherit them (in contrast to most _other_ deps where we _do_ tent to always list then), this PR makes that vote official and adds linting for it.

So in the general case we _do_ list things and _do not_ lint for missing ones -- for these two I propose that we _must not_ list things and _should_ lint for extra ones.

(3)

That's a very interesting point.  I think all of my reasons from (1) and (2) also apply to `:essential`.  I will try a prototype of this to see how it feels.

(4)

My hope was that a clear linter error that pointed users to in-tree documentation would be sufficient, without touching the styleguide.  (We don't really have a BUILD styleguide.  Maybe we should.) 

In my next push, I'll take a spin at clarifying this workflow.  I should also probably make only `--config lint` complain about this, instead of forbidding `bazel build` from working.  I think that'll make the error reporting easier, too.

**Comment by jwnimmer-tri** [comment]
The discussions here remain interesting, but are not as much of a pain point as previously, so I'm content to defer for now.
